crates/staged-sync/src/builder.rs

```diff
@@ -0,0 +1,317 @@
+    {
+        let fetch_client = Arc::new(self.network.fetch_client().await?);
+        Ok::<_, eyre::Error>(self.headers.as_ref().map(|config| {
+            let downloader = LinearDownloadBuilder::default()
+                .request_limit(config.downloader_batch_size)
+                .stream_batch_size(config.commit_threshold as usize)
+                // NOTE: the head and target will be set from inside the stage before the
+                // downloader is called
+                .build(
+                    self.consensus.clone(),
+                    fetch_client.clone(),
+                    Default::default(),
+                    Default::default(),
+                );
+
+            HeaderStage {
+                downloader,
+                consensus: self.consensus.clone(),
+                client: fetch_client.clone(),
+                network_handle: self.network.clone(),
+                metrics: HeaderMetrics::default(),
+            }
+        }))
+    }
+
+    /// Retrieves the [`TotalDifficultyStage`] if it was configured
+    pub fn total_difficulty_stage(&self) -> Option<TotalDifficultyStage> {
+        self.total_difficulty
+            .as_ref()
+            .map(|config| TotalDifficultyStage { commit_threshold: config.commit_threshold })
+    }
+
+    /// Returns the currently configured `BodyStage` if a `BodiesConfig` has been provided.
+    pub async fn bodies_stage(
+        &self,
+    ) -> Result<Option<BodyStage<ConcurrentDownloader<FetchClient, C>, C>>> {
+        let fetch_client = Arc::new(self.network.fetch_client().await?);
+        Ok::<_, eyre::Error>(self.bodies.as_ref().map(|config| {
+            let downloader =
+                ConcurrentDownloader::new(fetch_client.clone(), self.consensus.clone())
+                    .with_batch_size(config.downloader_batch_size)
+                    .with_retries(config.downloader_retries)
+                    .with_concurrency(config.downloader_concurrency);
+
+            BodyStage {
+                downloader: Arc::new(downloader),
+                consensus: self.consensus.clone(),
+                commit_threshold: config.commit_threshold,
+            }
+        }))
+    }
+
+    /// Given a [`Pipeline`] it proceeds to push the internally configured stages to it.
+    pub async fn configure_pipeline<DB: Database, U: SyncStateUpdater>(
+        &self,
+        mut pipeline: Pipeline<DB, U>,
+    ) -> Result<Pipeline<DB, U>> {
+        if let Some(stage) = self.headers_stage().await? {
+            pipeline = pipeline.push(stage);
+        }
+
+        if let Some(stage) = self.total_difficulty_stage() {
+            pipeline = pipeline.push(stage);
+        }
+
+        if let Some(stage) = self.bodies_stage().await? {
+            pipeline = pipeline.push(stage);
+        }
+
+        Ok(pipeline)
+    }
+
+    /// Builds and runs the pipeline
+    pub async fn run<DB: Database>(&self, db: Arc<DB>) -> Result<()> {
```

Comment:
this should probably consume the type make it impossible to configure everything multiple times